### PR TITLE
Add Impersonate-Uid description to Authentication docs page.

### PR DESCRIPTION
**Pull request by margocrawf**
This change goes with https://github.com/kubernetes/kubernetes/pull/99961
in the Kubernetes repo.

### CODE REVIEW

**Comment by sftim** [comment]
I think this change should target the docs release where that change will be made. If this will be part of the v1.22 release then @margocrawf please update this PR to target branch _dev-1.22_

(the primary branch of _this_ repo represents the current published website)

**Comment by sftim** [comment]
@PI-Victor FYI

**Comment by PI-Victor** [comment]
/assign
thanks @sftim

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:741]: Add a note like this for the new UID header.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:777]: Update `An example set of headers:` above with the uid header.

Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:777]: Above, update the `kubectl` example to point out that kubectl cannot impersonate extra and UID.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:811]: Update the `resourceNames` example below with a UID example.

**Review by enj — CHANGES_REQUESTED** [review]

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/authentication.md:738]: Should we explain what a UID is? We can use a tooltip for the glossary entry that covers it.
```suggestion
* `Impersonate-Uid`: The {{< glossary_tooltip text="UID" term_id="uid" >}} associated with the user being impersonated. Optional. Requires "Impersonate-User",
```

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/authentication.md:795]:
```suggestion
For impersonation, extra fields and impersonated UIDs are both under the
`authentication.k8s.io` API group.
```

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (margocrawf) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: That sounds good. I'm not familiar with the glossary tooltip, is that going to link to the uid entry in [this glossary?](https://kubernetes.io/docs/reference/glossary/?fundamental=true)

Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: Actually that'll link to https://k8s.io/docs/concepts/overview/working-with-objects/names (we could separately tweak that to link to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids)

If you hover over the text (“UID”) you see the first part of the glossary definition.

Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: Here's the glossary entry that the tooltip uses: https://github.com/kubernetes/website/blob/main/content/en/docs/reference/glossary/uid.md

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (margocrawf) [reply content/en/docs/reference/access-authn-authz/authentication.md:782]: Resolved, I just used a random hex string, but do you think I should do something that's more obviously a placeholder value (like `aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa` or `xxxxxxxx`)?

Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:782]: I think it's fine as-is. You could switch to something that's a valid UUID but still obvious, eg `123e4567-e89b-42d3-a456-​556642440000`

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/authentication.md:748]: Maybe this is the point to split it into 2 examples. The first example only sets `Impersonate-User:` and `Impersonate-Group:` headers; the second example sets  `Impersonate-User:` and every other header from the proposed example here, _except_ `Impersonate-Group:`

Motivation: make it easier to spot which headers are optional; reduce risk of any readers guessing that `Impersonate-UID:` is mandatory.

(it's fine to merge this as-is, I might propose that tweak separately once this merges).

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (margocrawf) [reply content/en/docs/reference/access-authn-authz/authentication.md:782]: Ok, if it's fine with you I think I'll just leave it as is.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (margocrawf) [reply content/en/docs/reference/access-authn-authz/authentication.md]: Resolved, I went ahead and split it.

**Comment by sftim** [comment]
Thanks @margocrawf 

[Preview](https://deploy-preview-28430--kubernetes-io-vnext-staging.netlify.app/docs/reference/access-authn-authz/authentication/) LGTM

@kubernetes/sig-auth-pr-reviews please tech review this one and add /lgtm if happy

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:789]:
```suggestion
`kubectl` cannot impersonate extra fields or UIDs.
```

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:738]: The tooltip reads:

> A Kubernetes systems-generated string to uniquely identify objects.

@sftim I find this confusing because you can impersonate any UID you want as long as you have the correct authorization.  This impersonated value is not Kubernetes generated and is not validated to be "correct" in any way.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:792]:
```suggestion
To impersonate a user, group, uid, or extra fields, the impersonating user must
```

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [inline content/en/docs/reference/access-authn-authz/authentication.md:809]:
```suggestion
For impersonation, extra fields and impersonated UIDs are both under the "authentication.k8s.io" `apiGroup`.
```

**Review by enj — CHANGES_REQUESTED** [review]
Minor stuff.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: Ah OK. I wonder we don't we call these UUIDs (like Wikipedia, etc do)? UID is widely used in POSIX to mean user id (contrast GID).

Anyway, given that it's not linked to Kubernetes' `metadata.uid` field, I suggest amending this to either:
```suggestion
* `Impersonate-Uid`: A UUID that represents or is associated with the user being
   impersonated. Optional. Requires "Impersonate-User".
```
or
```suggestion
* `Impersonate-Uid`: An opaque string that represents or is associated with the user being
   impersonated. Optional. Requires "Impersonate-User". Kubernetes does not impose any
   format requirements on this string other than that it is non-empty.
```
Without reading the code, I can't guess which is meant and I imagine other readers might also have trouble working that out.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:792]:
```suggestion
To impersonate a user, group, user UUID, or extra fields, the impersonating user must
```
or
```suggestion
To impersonate a user, group, user identifier (UID), or extra fields, the impersonating user must
```
or
```suggestion
To make any request involving impersonation, the user attempting impersonation must
```
(I'm not sure which is more appropriate)

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: Here are the [docs](https://github.com/kubernetes/kubernetes/blob/584658a17d958f0b663cb39dec1907cbba641300/staging/src/k8s.io/api/authentication/v1/types.go#L111-L115).  I am not sure if `A unique value that identifies this user across time.` is any more understandable for an end user though.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [reply content/en/docs/reference/access-authn-authz/authentication.md:792]: > To impersonate a user, group, user identifier (UID), or extra fields, the impersonating user must

That sounds the best to me.

File: content/en/docs/reference/access-authn-authz/authentication.md
Comment (enj) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: How about:

```
An unique identifier that represents the user being impersonated. Optional. Requires "Impersonate-User". Kubernetes does not impose any format requirements on this string.
```

The "other than that it is non-empty" bit is weird since the API server would just ignore the header (there is no distinction between unset and empty).

Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: I'm still confused. The API doc suggests that UIDs aren't reused, which makes it sound like it's either a CustomResourceDefinition or some other Kubernetes API. On the other hand, if the UID is exogenous (like the example `42`) then who knows whether the other system that generates these UIDs reuses them or not.

I have seen (and, for that matter, am typing this on an example of) operating systems that will reuse userids.

If Kubernetes is _setting an expectation_ that UID reuse is problematic in the context of impersonation, should we document **that**?

Also, I can't imagine ever designing a system where a user can have a UID of [empty string] but who knows what weird other integration someone somewhere might try to connect? Every other string value is valid though.
If that part is too much detail, fine.

Comment (sftim) [reply content/en/docs/reference/access-authn-authz/authentication.md:738]: I'm happy with
```suggestion
* A unique identifier that represents the user being impersonated. Optional. Requires
  "Impersonate-User". Kubernetes does not impose any format requirements on this string.
```
(not “an unique”, for semi-complicated reasons)

**Pull request merged**
